src/engine/sidechain/engineshoutcast.cpp

```diff
@@ -455,6 +496,45 @@ bool EngineShoutcast::serverConnect() {
 
 void EngineShoutcast::write(unsigned char *header, unsigned char *body,
                             int headerLen, int bodyLen) {
+    struct shoutcastCacheObject *l_SCacheObj = (struct shoutcastCacheObject *) malloc(sizeof(struct shoutcastCacheObject));
+
+    if (l_SCacheObj == NULL)
+    {
+       qDebug() << "EngineShoutcast::write: Can't allocate memory!";
+       return;
+    }
+
+    memset(l_SCacheObj, 0x00, sizeof(struct shoutcastCacheObject));
+    l_SCacheObj->header = (unsigned char *)malloc(headerLen);
+
+    if (l_SCacheObj->header == NULL) {
+       qDebug() << "EngineShoutcast::write: Can't allocate memory!";
+       free(l_SCacheObj);
+       return;
+    }
+
+    l_SCacheObj->body = (unsigned char *)malloc(bodyLen);
+
+    if (l_SCacheObj->body == NULL) {
+       qDebug() << "EngineShoutcast::write: Can't allocate memory!";
+       free(l_SCacheObj->header);
+       free(l_SCacheObj);
+       return;
+    }
+
+    memcpy(l_SCacheObj->header, header, headerLen);
+    memcpy(l_SCacheObj->body, body, bodyLen);
+
+    l_SCacheObj->headerLen = headerLen;
+    l_SCacheObj->bodyLen = bodyLen;
+
+    m_SMutex.lock();
+    m_pShoutcastCache.append(l_SCacheObj);
```

Comment:
He plain C is more readable many times than some C++ files. Yes I looked FIFO for one option but the problem is that this compressed audio which makes things harder than having FIFO pipe and also this libshout way of separate header and body sending. So even using FIFO we got to store packet sizes or send all we got and hope for the best. I tried my best that those mutexes are used as short as they can. But if I can do this with FIFO I'll be happy puppy but I just don't understand how.